lua/weapons/weapon_laserpointer.lua

```diff
@@ -0,0 +1,599 @@
+		surface.SetDrawColor( ColorAlpha(NamedColor("FgColor"),alpha) )
+		surface.SetMaterial( wepicon )
+
+		local fsin = 0
+		local cx,cy = x+wide/2, y+tall/2.25	
+		local size = 168
+		local ix,iy = cx-size/2,cy-size/2	
+		surface.DrawTexturedRect( ix , iy ,  size ,  size )
+		self:PrintWeaponInfo( x + wide + 20, y + tall * 0.95, alpha )
+	end
+	
+	
+end
+
+local maxsat = 0.95
+SWEP.ColorFunctions = {}
+SWEP.ColorFunctions[0] = function(ply) 
+	local wpncolor = ply:GetWeaponColor()
+	local ch,cs,cv = ColorToHSV(Color(wpncolor.r*255,wpncolor.g*255,wpncolor.b*255))
+			cv = 1
+			cs = math.Clamp(cs*maxsat,0,maxsat)
+		local lasercolor = HSVToColor(ch,cs,cv)
+	return lasercolor
+end
+
+SWEP.ColorFunctions[1] = function(ply) 
+	local plycolor = ply:GetPlayerColor()
+	local ch,cs,cv = ColorToHSV(Color(plycolor.r*255,plycolor.g*255,plycolor.b*255))
+			cv = 1
+			cs = math.Clamp(cs*maxsat,0,maxsat)
+		local lasercolor = HSVToColor(ch,cs,cv)
+	return lasercolor
+end
+
+local function wtfhsv(h,s,v) --apparently HSVToColor does not create colors with a metatable so here's a shitty override function
+local c = HSVToColor(h,s,v)
+return Color(c.r,c.g,c.b)
+end
+
+SWEP.ColorFunctions[2] = wtfhsv(0,maxsat,1)
+SWEP.ColorFunctions[3] = wtfhsv(120,maxsat,1)
+SWEP.ColorFunctions[4] = wtfhsv(240,maxsat,1)
+
+SWEP.ColorFunctions[5] = wtfhsv(60,maxsat,1)
+SWEP.ColorFunctions[6] = wtfhsv(180,maxsat,1)
+SWEP.ColorFunctions[7] = wtfhsv(300,maxsat,1)
+
+SWEP.ColorFunctions[8] = function(ply) 
+	local plycolor = ply:GetPlayerColor()
+	ply.lzroffset = ply.lzroffset or math.Rand(0,360)
+	local lasercolor = HSVToColor(180+math.NormalizeAngle(ply.lzroffset + CurTime()*720),maxsat,1)
+	return lasercolor
+end
+
+
+
+
+function SWEP:ButtonSound(state)
+local clicksound = self.ClickSound
+if(!state)then clicksound = self.UnClickSound end
+self:EmitSound(clicksound,10,150,0.001)
+end
+
+function SWEP:Initialize()
+	if(CLIENT and IsValid(self:GetOwner()))then self.ViewModelFOV = self:GetOwner():GetFOV() end
+	self:SetHoldType( "pistol" )
+end
+
+function SWEP:SetupDataTables()
+	self:NetworkVar( "Bool", 0, "OnState" )
+	self:NetworkVar( "Bool", 1, "BeamMode" )
+	self:NetworkVar( "Int", 0, "ColorMode" )
+end
+
+
+hook.Add("NetworkEntityCreated","LaserPointerFind",function(ent) 
+	if ent:GetClass() == "weapon_laserpointer" then
+	lpointer_laser_source[ent] = true 
+	end
+end)
+
+hook.Add("EntityRemoved","LaserPointerRemove",function(ent) 
+	if ent:GetClass() == "weapon_laserpointer" then
+	lpointer_laser_source[ent] = nil 
+	end
+end)
+
+function SWEP:SecondaryAttack() 
+	local can = hook.Call("PlayerCanHaveLaserBeams",self:GetOwner(),self)
+	if(can != false)then 
+	self:SetBeamMode(!self:GetBeamMode())
+	self:ButtonSound(true)
+	end
+	self:SetNextSecondaryFire(CurTime() + 0.2)
+end
+
+function SWEP:Deploy()
+	if(CLIENT)then self.HornBG = nil end
+	if(CLIENT and IsValid(self:GetOwner()))then self.ViewModelFOV = self:GetOwner():GetFOV() end
```

Comment:
i think this breaks if you press suit zoom key or change the setting, can you use a SWEP:SetupView hook or whatever it is?

Reply:
The viewmodel FOV needs to match the camera's current fov or the position of the laser beam in the viewmodel is misaligned. You can't even shoot the laser while suit zoom is active

Reply:
I understand why, I mean you can continuously update this value in SetupView instead of just once

Reply:
https://wiki.facepunch.com/gmod/WEAPON:CalcView just put the fov argument into the class here